### PR TITLE
Max flow implementation with increased functionality and LCA improvements supporting weighted edges

### PR DESCRIPTION
**Pull request by cameroncuster**
#  New Maxflow Struct
The new maxflow implementation is described along with complexity in the source. The implementation supports increased functionality to the original Dinic implementation, but at the cost of slower run time.

## Utility
- find edges in the minimum cut
- flow of single edge
- send single flow across edge
- supports Dinic and Edmonds-Karp

## Results and Cross Comparisons
- the original maxflow struct was about 15x faster on this problem [Maximum Flow](https://open.kattis.com/problems/maxflow) where the only details modified where the struct used in the maxflow calculations
- Original imp (0.02s)
- New imp (0.3s)

image
Here are the cross comparisons of the flow implementations. Ignore the WA. The first one is original Dinic. The newest one is Edmonds-Karp with new struct, and the second to latest one is Dinic with the new struct.

## Testing
- The new struct is substantially slower, but has been tested on a variety of problems and proven it's utility
- Here are a few problems I have used for testing the new imp: [Distinct Routes](https://cses.fi/problemset/task/1711) - [sol](https://github.com/cameroncuster/CSESProblemSet/blob/main/graph_algorithms/distinct_routes.cpp), [Police Chase](https://cses.fi/problemset/task/1695) - [sol](https://github.com/cameroncuster/CSESProblemSet/blob/main/graph_algorithms/police_chase.cpp), [Download Speed](https://cses.fi/problemset/task/1694) - [sol](https://github.com/cameroncuster/CSESProblemSet/blob/main/graph_algorithms/download_speed.cpp)

# LCA Improvements and Utility
- supports weighted trees
- supports distance in edges and weight of path operations using LCA
- smaller include

## Testing
- Tested with this problem [The Shortest Statement](https://codeforces.com/contest/1051/problem/F) - [sol](https://codeforces.com/contest/1051/submission/124378359)

### CODE REVIEW

**Comment by cameroncuster** [comment]
image
Here are the cross comparisons of the flow implementations. Ignore the WA. The first one is original Dinic. The newest one is Edmonds-Karp with new struct, and the second to latest one is Dinic with the new struct.

File: graphs/lca.cpp [hunk 1]
```diff
@@ -1,53 +1,65 @@
...
-		dfs(root,root,graph);
-	}
+	vector<int> depth, dist;
 
-	void dfs(int node, int par, const vector<vector<int>> &graph) {
+	// use weights of 1 for unweighted tree
+	lca(const vector<vector<pair<int, ll>>> &graph, int root) :
+		depth(graph.size()), dist(graph.size()) {//0 - based nodes
+			int n = graph.size();
+			Log = 1;
```
Comment (lrvideckis) [inline graphs/lca.cpp:20]: initializer list

File: graphs/lca.cpp [hunk 2]
```diff
@@ -1,53 +1,65 @@
...
-		for(int to : graph[node]) {
+		for(auto [to, w] : graph[node]) {
 			if(to == par) continue;
 			depth[to] = 1 + depth[node];
+			dist[to] = w + dist[node];
 			dfs(to, node, graph);
 		}
 	}
 
 	int kthPar(int node, int k) {
```
Comment (lrvideckis) [inline graphs/lca.cpp:38]: make const

File: graphs/lca.cpp [hunk 3]
```diff
@@ -1,53 +1,65 @@
...
-			if(k&(1<<bit)) {
+		for(int bit = 0; bit < Log; ++bit)
+			if(k&(1<<bit))
 				node = memo[node][bit];
-			}
-		}
 		return node;
 	}
 
 	int getLca(int x, int y) {
```
Comment (lrvideckis) [inline graphs/lca.cpp:45]: make const while were here

File: graphs/lca.cpp [hunk 4]
```diff
@@ -1,53 +1,65 @@
...
-
-	lca(const vector<vector<int>> &graph, int root) {//0 - based nodes
-		int n = graph.size();
-		Log = 1;
-		while((1<<Log) < n) ++Log;
-		depth.resize(n,0);
-		memo.resize(n,vector<int>(Log));
-		dfs(root,root,graph);
-	}
+	vector<int> depth, dist;
```
Comment (lrvideckis) [inline graphs/lca.cpp:14]: dist should be a ll. Can you test with a problem where longs e required?

Comment (cameroncuster) [reply graphs/lca.cpp:14]: Yeah I see, I don't really know one, do you?

**Comment by lrvideckis** [comment]
I’m against adding a slower max flow implementation. LMK if you want help speeding it up

File: graphs/lca.cpp (hunk 3 above)
Comment (cameroncuster) [reply graphs/lca.cpp:44]: https://codeforces.com/contest/1051/submission/124398011

File: graphs/lca.cpp (hunk 2 above)
Comment (cameroncuster) [reply graphs/lca.cpp:37]: done - submission below

File: graphs/lca.cpp (hunk 1 above)
Comment (cameroncuster) [reply graphs/lca.cpp]: For Log? - ammended - see submission below

**Review by lrvideckis — CHANGES_REQUESTED** [review]
Please address the new comments

**Comment by cameroncuster** [comment]
I do not know of a problem requiring longs that I can use to test it... otherwise all of your requests have been fulfilled for the LCA to my understanding.

File: graphs/lca.cpp (hunk 4 above)
Comment (lrvideckis) [reply graphs/lca.cpp:14]: Hmm, couldn't find one. Can you test instead by writing naive, then running this against naive on random test data that requires longs?

Comment (lrvideckis) [reply graphs/lca.cpp:14]: And send me the files or results somehow

Comment (cameroncuster) [reply graphs/lca.cpp:14]: Yeah that'd work, might do this in the morning as I am in the midst of a random Web dev project

File: graphs/lca.cpp [hunk 5]
```diff
@@ -1,53 +1,64 @@
...
 
-	void dfs(int node, int par, const vector<vector<int>> &graph) {
+	// use weights of 1 for unweighted tree
```
Comment (lrvideckis) [inline graphs/lca.cpp:16]: Let's remove as we still have dist edges function

Comment (cameroncuster) [reply graphs/lca.cpp:16]: Remove the DFS?

File: graphs/lca.cpp (hunk 4 above)
Comment (lrvideckis) [reply graphs/lca.cpp:14]: lol this problem should work https://codeforces.com/contest/1051/problem/F

Comment (lrvideckis) [reply graphs/lca.cpp:14]: (Don't forget to remove the #define int long long when testing

Comment (cameroncuster) [reply graphs/lca.cpp:14]: ah that makes sense. I was running with the macro so I guess that's why it worked.

File: graphs/lca.cpp (hunk 5 above)
Comment (lrvideckis) [reply graphs/lca.cpp:16]: remove the comment

Comment (lrvideckis) [reply graphs/lca.cpp:16]: actually nevermind, let's keep it

File: graphs/lca.cpp (hunk 2 above)
Comment (lrvideckis) [reply graphs/lca.cpp:37]: I meant make the function const:
int kthPar(...) const {

**Review by lrvideckis — CHANGES_REQUESTED** [review]
new comments

File: graphs/lca.cpp (hunk 5 above)
Comment (cameroncuster) [reply graphs/lca.cpp:16]: okay sounds good

File: graphs/lca.cpp (hunk 2 above)
Comment (cameroncuster) [reply graphs/lca.cpp:37]: okay sounds good

File: graphs/lca.cpp (hunk 4 above)
Comment (cameroncuster) [reply graphs/lca.cpp]: https://codeforces.com/contest/1051/submission/124627472 here is the passing test. Can you make sure the code updated, I think it just pulls latest commit from my master until merged?

**Comment by cameroncuster** [comment]
Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?

**Comment by lrvideckis** [comment]
> Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?

sure, after comments are addressed

**Comment by cameroncuster** [comment]
> > Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?
> 
> sure, after comments are addressed

Have I not addressed all comments? Did the source not update on your end?

**Comment by lrvideckis** [comment]
> > > Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?
> > 
> > sure, after comments are addressed
> 
> Have I not addressed all comments? Did the source not update on your end?

- make functions const
- comment link of submission testing for ll’s

**Comment by cameroncuster** [comment]
> > > > Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?
> > > 
> > > 
> > > sure, after comments are addressed
> > 
> > 
> > Have I not addressed all comments? Did the source not update on your end?
> 
> * make functions const
> * comment link of submission testing for ll’s

I commented the link as a reply above ^,
and I made the functions const (really only two can be)

**Comment by cameroncuster** [comment]
Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795

**Comment by cameroncuster** [comment]
> > > > Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?
> > > 
> > > 
> > > sure, after comments are addressed
> > 
> > 
> > Have I not addressed all comments? Did the source not update on your end?
> 
> * make functions const
> * comment link of submission testing for ll’s

https://github.com/lrvideckis/Programming-Team-Code/pull/6#discussion_r680595236 here is a link to the comment containing the functions tested with longs

**Comment by lrvideckis** [comment]
> Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795

you have define int ll

**Comment by lrvideckis** [comment]
but I saw the other submission, so we’re good

**Comment by cameroncuster** [comment]
> > Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795
> 
> you have define int ll

Yeah I know, I already did the other one, and this was easier.

**Comment by cameroncuster** [comment]
> Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795

Wait why did you 'dislike' this one? Because of the int to long long macro?

**Comment by lrvideckis** [comment]
> > Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795
> 
> Wait why did you 'dislike' this one? Because of the int to long long macro?

exactly - It doesn’t test for longs cuz of the macro

**Comment by cameroncuster** [comment]
> > > Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795
> > 
> > 
> > Wait why did you 'dislike' this one? Because of the int to long long macro?
> 
> exactly - It doesn’t test for longs cuz of the macro

Just to make sure we are keeping our testing standards the highest https://codeforces.com/contest/1051/submission/124637812

**Comment by cameroncuster** [comment]
Thankfully I'm really like the LCA implementation we have now. I think it offers decent functionality for any problem I have seen (which are few)

File: rangeDS/fenwickTree.cpp [hunk 6]
```diff
@@ -1,6 +1,16 @@
 struct fenwickTree {
 	vector<ll> bit;
 	fenwickTree(int n) : bit(n,0) {}
+	fenwickTree(const vector<int> & a) : bit(a.size()) {
```
Comment (lrvideckis) [inline rangeDS/fenwickTree.cpp:4]: Vector of ll

File: rangeDS/fenwickTree.cpp [hunk 7]
```diff
@@ -1,6 +1,16 @@
...
+		bit[0] = a[0];
+		for (int i = 1; i < (int)a.size(); i++)
+			bit[i] = bit[i-1] + a[i];
+		for (int i = a.size() - 1; i > 0; i--) {
```
Comment (lrvideckis) [inline rangeDS/fenwickTree.cpp:8]: Cast a.size to an int, otherwise this will underflow in the case where a is empty

Comment (lrvideckis) [reply rangeDS/fenwickTree.cpp:8]: Then you'll get essentially a runtime error, it's a pretty nasty bug

**Review by lrvideckis — CHANGES_REQUESTED** [review]

File: rangeDS/fenwickTree.cpp [hunk 8]
```diff
@@ -1,6 +1,16 @@
 struct fenwickTree {
 	vector<ll> bit;
 	fenwickTree(int n) : bit(n,0) {}
+	fenwickTree(const vector<int> & a) : bit(a.size()) {
+		bit[0] = a[0];
```
Comment (lrvideckis) [inline rangeDS/fenwickTree.cpp:5]: This will RTE if a is empty

**Pull request closed without merging**